plutus-core/plutus-core/src/PlutusCore/Name/UniqueSet.hs

```diff
@@ -0,0 +1,91 @@
+module PlutusCore.Name.UniqueSet (
+  UniqueSet (..),
+  insertByUnique,
+  insertByName,
+  singletonName,
+  fromFoldable,
+  fromUniques,
+  fromNames,
+  memberByUnique,
+  memberByName,
+  notMemberByName,
+  (\\),
+  union,
+  setOfByUnique,
+  setOfByName,
+) where
+
+import Control.Lens (Getting, view)
+import Control.Lens.Getter (views)
+import Data.Coerce (Coercible, coerce)
+import Data.Foldable (foldl')
+import Data.IntSet qualified as IS
+import Data.IntSet.Lens qualified as IS
+import PlutusCore.Name.Unique (HasUnique (..), Unique (Unique))
+
+-- | A set containing uniques.
+newtype UniqueSet unique = UniqueSet
+  { unUniqueSet :: IS.IntSet
+  }
+  deriving newtype (Show, Eq, Semigroup, Monoid)
+
+-- | Insert a value by a unique.
+insertByUnique ::
+  (Coercible unique Unique) =>
+  unique ->
+  UniqueSet unique ->
+  UniqueSet unique
+insertByUnique = coerce . IS.insert . coerce
+
+-- | Insert a value by the unique of a name.
+insertByName :: (HasUnique name unique) => name -> UniqueSet unique -> UniqueSet unique
+insertByName = insertByUnique . view unique
```

Comment:
tbh, I slightly question whether these functions pass the Fairbarn threshold?

Reply:
I realise most of them existed already, I'm just questioning whether they should!

Reply:
Other functions are defined based on `insertByUnique` and `insertByName` to avoid redoing all back and forth between the ints and the uniques and the names